### PR TITLE
Add Async execution helper

### PR DESCRIPTION
**Pull request by daniel-dressler**
This async execution helper assists with deep callback nesting. It may not solve the core issue of deep nested callbacks but atleast it makes reading possible.

This module has no dependencies and is thus being PR'd async with the signletons. It should merge clean, I hope.

### CODE REVIEW

File: Singleton/MonoSingleton.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using UnityEngine;
+
+public class MonoSingleton<T> : MonoBehaviour where T : class, new() {
+	// Instance functions
+	private static T _Instance;
```
Comment (nullorvoid) [inline Singleton/MonoSingleton.cs:5]: Should match case with Singleton.cs

File: Async/Async.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			fn(items[currentItemI], (Exception error) => {
+				// Stop iteration if there was an error
+				if (error != null) {
+					cb(error);
+					return;
+				}
+
+				// Continue to next item
+				currentItemI++;
+				iterate();
```
Comment (ronkorving) [inline Async/Async.cs:37]: To avoid massive stack traces on long lists of items, should we use Coroutines?

File: Async/Async.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+			if (currentItemI >= actionItems.Count) {
+				cb(null);
+				return;
+			}
+			
+			// Shift an element from the list
+			Action<Action<Exception>> actionItem = actionItems[currentItemI];
+			
+			// Execute the given function on this item
+			actionItem((Exception error) => {
```
Comment (ronkorving) [inline Async/Async.cs:64]: The callback you pass here creates yet another closure, which I don't think needs to exist.  This function could be instantiated at the same level as your function that starts at line 54, no? That should save you memory and garbage.

File: Async/Async.cs (hunk 2 above)
Comment (nullorvoid) [inline Async/Async.cs:37]: By Coroutines you mean making this library class depend on Unity?  I'm not sure thats a good idea, if it's even possible? (main thread problems?)

Daniel also suggests that as a developer you may want the full stack for this as it is a series of function calls by the user.

File: Async/Async.cs (hunk 3 above)
Comment (nullorvoid) [inline Async/Async.cs:64]: We tested the suggestion, seems to work fine moving this outside. 

```
(Exception error) => 
{
    // Stop iteration if there was an error
    if (error != null) {
        cb (error);
        return;
    }

    // Continue to next item
    currentItemI++;
    iterate ();
};
```

We also think that the names here are pretty nondescript so we will like to name these properly.

**Comment by daniel-dressler** [comment]
@ronkorving We've gone over this PR and pulled out the issues. We're good to go.

LGTM

File: Async/Async.cs (hunk 2 above)
Comment (ronkorving) [inline Async/Async.cs:37]: Coroutines are not a C# feature?
You may want the full stack, until you hit a stack overflow and you're simply screwed. This _does_ happen, and there will be nothing you can do about it. Besides, you're supposed to be doing asynchronous work anyway. What makes you think there will be a "full" stack trace? (in which case, ironically I guess we're fine without coroutines).

File: Async/Async.cs (hunk 3 above)
Comment (ronkorving) [inline Async/Async.cs:64]: So can you make the move of that function final?

File: Async/Async.cs (hunk 2 above)
Comment (nullorvoid) [inline Async/Async.cs:37]: Coroutines are kinda in C# (emulated with yield but this isn't full coroutines there are some restrictions, you could also use threading i guess) but the "Coroutine" function is provided in Unity.Engine.

File: Async/Async.cs (hunk 3 above)
Comment (daniel-dressler) [inline Async/Async.cs:64]: Sorry Ron for these PRs we are gathering the issues. Refactoring will come once we have everything.

**Pull request merged**